pennylane/qnodes/decorator.py

```diff
@@ -21,20 +21,140 @@
+        diff_method (str, None): the method of differentiation to use in the created QNode
+
+    Raises:
+        ValueError: if an unrecognized ``diff_method`` is provided
+
+    Returns:
+        ~.BaseQNode: the QNode class object that is compatible with the provided device and
+        differentiation method
+    """
+    # pylint: disable=too-many-return-statements,too-many-branches
+    model = device.capabilities().get("model", "qubit")
+    allows_passthru = device.capabilities().get("passthru", False)
+    device_provides_jacobian = device.capabilities().get("provides_jacobian", False)
+
+    if allows_passthru:
+        passthru_interface = device.capabilities()["passthru_interface"]
+
+    if diff_method is None:
+        # QNode is not differentiable
+        return BaseQNode
+
+    if diff_method == "best":
+
+        if allows_passthru:
+            if interface == passthru_interface:
+                # hand off differentiation to the device without type conversion
+                return PassthruQNode
+
+        if device_provides_jacobian:
+            # hand off differentiation to the device
+            return DeviceJacobianQNode
+
+        if model in PARAMETER_SHIFT_QNODES:
+            # parameter-shift analytic differentiation
+            return PARAMETER_SHIFT_QNODES[model]
+
+    if diff_method == "classical":
+        if allows_passthru:
+            if interface != passthru_interface:
+                raise ValueError(
+                    "Device {} only supports the {} interface when "
+                    "diff_method='classical'".format(device.short_name, passthru_interface)
+                )
+            return PassthruQNode
+
+        raise ValueError(
+            "The {} device does not support native computations with "
+            "autodifferentiation frameworks.".format(device.short_name)
+        )
+
+    if diff_method == "device":
+        if device_provides_jacobian:
+            return DeviceJacobianQNode
+
+        raise ValueError(
+            "The {} device does not provide a native method "
+            "for computing the jacobian.".format(device.short_name)
+        )
+
+    if diff_method == "parameter-shift":
+        if model in PARAMETER_SHIFT_QNODES:
+            # parameter-shift analytic differentiation
+            return PARAMETER_SHIFT_QNODES[model]
+
+        raise ValueError(
+            "The {} device does not provide a native method "
+            "for computing the jacobian.".format(device.short_name)
+        )
+
+    if diff_method in ALLOWED_DIFF_METHODS:
+        # finite differences
+        return JacobianQNode
+
+    raise ValueError(
+        "Differentiation method {} not recognized. Allowed "
+        "options are {}".format(diff_method, ALLOWED_DIFF_METHODS)
+    )
+
+
+def _apply_interface(qnode_, interface, diff_method):
+    """Applies an interface to a specified QNode.
+
+    Args:
+        qnode_ (~.BaseQNode): the QNode to which the interface is applied
```

Comment:
Just curious: any significance in having the post underscore in the name? (`qnode_`)

Reply:
This was because the linter was complaining that `qnode` was being re-defined from the outer scope (since it was being used for both the top-level decorator function, as well as a local variable within this function).

I wasn't sure what else to name it to, so appended an underscore! Since this is a private function, and not user facing, the argument name doesn't matter too much. I've also seen other people append an underscore to avoid clashing variable names 😆 

Feel free to change it!

Reply:
Nope, totally fine with me! :)